File: src/services/FISCardService.ts
/**
 * FIS Global Payment One - Card Management Service
 *
 * Handles all FIS card operations for the SwipeSavvy mobile app:
 * - Card issuance (virtual/physical)
 * - Card controls (lock, limits, channels)
 * - PIN management
 * - Transactions
 * - Digital wallet provisioning
 * - Fraud alerts
 */

import { dataService } from './DataService';

// =============================================================================
// TYPES
// =============================================================================

export interface FISCard {
  id: string;
  fisCardId: string;
  cardType: 'virtual' | 'physical';
  status: 'pending' | 'active' | 'locked' | 'frozen' | 'cancelled' | 'expired';
  lastFour: string;
  expiryMonth: number;
  expiryYear: number;
  cardholderName: string;
  cardNetwork: 'visa' | 'mastercard';
  nickname?: string;
  isPrimary: boolean;
  pinSet: boolean;
  pinLocked: boolean;
  createdAt: string;
  activatedAt?: string;
  // Shipping for physical cards
  shippingStatus?: 'processing' | 'shipped' | 'in_transit' | 'delivered';
  trackingNumber?: string;
  shippingCarrier?: string;
}

export interface FISCardControls {
  cardId: string;
  // Spending limits
  dailyLimit?: number;
  weeklyLimit?: number;
  monthlyLimit?: number;
  perTransactionLimit?: number;
  // Channel controls
  atmEnabled: boolean;
  posEnabled: boolean;
  ecommerceEnabled: boolean;
  contactlessEnabled: boolean;
  internationalEnabled: boolean;
  // Merchant controls
  blockedMccCodes: string[];
  allowedMccCodes?: string[];
  // Geographic controls
  allowedCountries: string[];
  blockedCountries: string[];
  // Alert preferences
  alertOnTransaction: boolean;
  alertOnDecline: boolean;
  alertOnInternational: boolean;
  alertThreshold?: number;
}

export interface FISTransaction {
  id: string;
  cardId: string;
  transactionType: 'purchase' | 'refund' | 'atm_withdrawal' | 'atm_deposit' | 'transfer' | 'fee';
  status: 'pending' | 'posted' | 'declined' | 'reversed' | 'disputed';
  amount: number;
  currency: string;
  merchantName?: string;
  merchantId?: string;
  mccCode?: string;
  mccDescription?: string;
  category?: string;
  channel: 'pos' | 'atm' | 'ecommerce' | 'contactless' | 'mobile';
  merchantCity?: string;
  merchantState?: string;
  merchantCountry?: string;
  isInternational: boolean;
  rewardsEarned?: number;
  notes?: string;
  timestamp: string;
  postedAt?: string;
}

export interface FISAlert {
  id: string;
  cardId: string;
  alertType: 'large_transaction' | 'international_transaction' | 'declined_transaction' | 'suspicious_activity' | 'pin_attempt_failed';
  priority: 'low' | 'medium' | 'high' | 'critical';
  status: 'new' | 'acknowledged' | 'resolved';
  title: string;
  description: string;
  amount?: number;
  merchantName?: string;
  createdAt: string;
}

export interface FISWalletToken {
  id: string;
  cardId: string;
  walletType: 'apple_pay' | 'google_pay' | 'samsung_pay';
  status: 'requested' | 'active' | 'suspended' | 'deleted';
  deviceId: string;
  deviceType: 'phone' | 'watch' | 'tablet';
  deviceName?: string;
  tokenSuffix: string;
  createdAt: string;
  lastUsedAt?: string;
}

export interface TravelNotice {
  id: string;
  cardId: string;
  startDate: string;
  endDate: string;
  destinations: string[];
  notes?: string;
  createdAt: string;
}

export interface IssueVirtualCardRequest {
  cardholderName: string;
  nickname?: string;
  setAsPrimary?: boolean;
}

export interface IssuePhysicalCardRequest {
  cardholderName: string;
  shippingAddress: {
    street: string;
    city: string;
    state: string;
    zip: string;
    country?: string;
  };
  expedited?: boolean;
  nickname?: string;
  setAsPrimary?: boolean;
}

export interface SpendingLimits {
  dailyLimit?: number;
  weeklyLimit?: number;
  monthlyLimit?: number;
  perTransactionLimit?: number;
}

export interface ChannelControls {
  atmEnabled?: boolean;
  posEnabled?: boolean;
  ecommerceEnabled?: boolean;
  contactlessEnabled?: boolean;
  internationalEnabled?: boolean;
}

// =============================================================================
// FIS CARD SERVICE
// =============================================================================

const API_BASE = 'https://api.swipesavvy.com/api/v1';

class FISCardService {
  private getHeaders(): HeadersInit {
    const token = dataService.getAuthToken();
    return {
      'Content-Type': 'application/json',
      ...(token ? { 'Authorization': `Bearer ${token}` } : {}),
    };
  }

  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<T> {
    const url = `${API_BASE}${endpoint}`;

    try {
      const response = await fetch(url, {
        ...options,
        headers: {
          ...this.getHeaders(),
          ...options.headers,
        },
      });

      const data = await response.json();

      if (!response.ok) {
        throw new Error(data.detail || data.message || 'Request failed');
      }

      return data.data || data;
    } catch (error) {
      console.error(`FIS API Error [${endpoint}]:`, error);
      throw error;
    }
  }

  // ===========================================================================
  // CARD ISSUANCE & LIFECYCLE
  // ===========================================================================

  /**
   * Get all FIS cards for the current user
   */
  async getCards(): Promise<FISCard[]> {
    return this.request<FISCard[]>('/cards');
  }

  /**
   * Get a specific card by ID
   */
  async getCard(cardId: string): Promise<FISCard> {
    return this.request<FISCard>(`/cards/${cardId}`);
  }

  /**
   * Issue a new virtual card (instant)
   */
  async issueVirtualCard(request: IssueVirtualCardRequest): Promise<FISCard> {
    return this.request<FISCard>('/cards/issue/virtual', {
      method: 'POST',
      body: JSON.stringify({
        cardholder_name: request.cardholderName,
        nickname: request.nickname,
        set_as_primary: request.setAsPrimary || false,
      }),
    });
  }

  /**
   * Order a physical card (shipped)
   */
  async issuePhysicalCard(request: IssuePhysicalCardRequest): Promise<FISCard> {
    return this.request<FISCard>('/cards/issue/physical', {
      method: 'POST',
      body: JSON.stringify({
        cardholder_name: request.cardholderName,
        shipping_address: request.shippingAddress,
        expedited: request.expedited || false,
        nickname: request.nickname,
        set_as_primary: request.setAsPrimary || false,
      }),
    });
  }

  /**
   * Activate a card
   */
  async activateCard(cardId: string, lastFour: string, activationCode?: string): Promise<void> {
    await this.request(`/cards/${cardId}/activate`, {
      method: 'POST',
      body: JSON.stringify({
        last_four: lastFour,
        activation_code: activationCode,
      }),
    });
  }

  /**
   * Replace a card (lost, stolen, damaged)
   */
  async replaceCard(
    cardId: string,
    reason: 'lost' | 'stolen' | 'damaged' | 'expired',
    shippingAddress?: IssuePhysicalCardRequest['shippingAddress'],
    expedited?: boolean
  ): Promise<FISCard> {
    return this.request<FISCard>(`/cards/${cardId}/replace`, {
      method: 'POST',
      body: JSON.stringify({
        reason,
        shipping_address: shippingAddress,
        expedited: expedited || false,
      }),
    });
  }

  /**
   * Cancel a card permanently
   */
  async cancelCard(cardId: string, reason: string): Promise<void> {
    await this.request(`/cards/${cardId}`, {
      method: 'DELETE',
      body: JSON.stringify({ reason }),
    });
  }

  /**
   * Get shipping status for a physical card
   */
  async getShippingStatus(cardId: string): Promise<{
    status: string;
    trackingNumber?: string;
    carrier?: string;
    estimatedDelivery?: string;
  }> {
    return this.request(`/cards/${cardId}/shipping`);
  }

  /**
   * Get sensitive card data (PAN, CVV) - requires PIN verification
   */
  async getSensitiveData(
    cardId: string,
    includePan?: boolean,
    includeCvv?: boolean
  ): Promise<{ pan?: string; cvv?: string; expiry?: string }> {
    const params = new URLSearchParams();
    if (includePan) params.append('include_pan', 'true');
    if (includeCvv) params.append('include_cvv', 'true');

    return this.request(`/cards/${cardId}/sensitive?${params.toString()}`);
  }

  // ===========================================================================
  // CARD CONTROLS
  // ===========================================================================

  /**
   * Lock a card temporarily
   */
  async lockCard(cardId: string, reason?: string): Promise<void> {
    await this.request(`/cards/${cardId}/lock`, {
      method: 'POST',
      body: JSON.stringify({ reason }),
    });
  }

  /**
   * Unlock a card
   */
  async unlockCard(cardId: string): Promise<void> {
    await this.request(`/cards/${cardId}/unlock`, {
      method: 'POST',
    });
  }

  /**
   * Freeze a card (for suspected fraud)
   */
  async freezeCard(cardId: string, reason?: string): Promise<void> {
    await this.request(`/cards/${cardId}/freeze`, {
      method: 'POST',
      body: JSON.stringify({ reason }),
    });
  }

  /**
   * Unfreeze a card
   */
  async unfreezeCard(cardId: string): Promise<void> {
    await this.request(`/cards/${cardId}/unfreeze`, {
      method: 'POST',
    });
  }

  /**
   * Get all card controls
   */
  async getCardControls(cardId: string): Promise<FISCardControls> {
    return this.request<FISCardControls>(`/cards/${cardId}/controls`);
  }

  /**
   * Set spending limits
   */
  async setSpendingLimits(cardId: string, limits: SpendingLimits): Promise<void> {
    await this.request(`/cards/${cardId}/limits`, {
      method: 'PUT',
      body: JSON.stringify({
        daily_limit: limits.dailyLimit,
        weekly_limit: limits.weeklyLimit,
        monthly_limit: limits.monthlyLimit,
        per_transaction_limit: limits.perTransactionLimit,
      }),
    });
  }

  /**
   * Get spending limits
   */
  async getSpendingLimits(cardId: string): Promise<SpendingLimits> {
    return this.request<SpendingLimits>(`/cards/${cardId}/limits`);
  }

  /**
   * Remove all spending limits
   */
  async removeSpendingLimits(cardId: string): Promise<void> {
    await this.request(`/cards/${cardId}/limits`, {
      method: 'DELETE',
    });
  }

  /**
   * Set channel controls (ATM, POS, eCommerce, etc.)
   */
  async setChannelControls(cardId: string, controls: ChannelControls): Promise<void> {
    await this.request(`/cards/${cardId}/controls/channels`, {
      method: 'PUT',
      body: JSON.stringify({
        atm_enabled: controls.atmEnabled,
        pos_enabled: controls.posEnabled,
        ecommerce_enabled: controls.ecommerceEnabled,
        contactless_enabled: controls.contactlessEnabled,
        international_enabled: controls.internationalEnabled,
      }),
    });
  }

  /**
   * Enable international transactions
   */
  async enableInternational(cardId: string): Promise<void> {
    await this.request(`/cards/${cardId}/controls/international/enable`, {
      method: 'POST',
    });
  }

  /**
   * Disable international transactions
   */
  async disableInternational(cardId: string): Promise<void> {
    await this.request(`/cards/${cardId}/controls/international/disable`, {
      method: 'POST',
    });
  }

  /**
   * Block a merchant category
   */
  async blockMerchantCategory(cardId: string, category: string): Promise<void> {
    await this.request(`/cards/${cardId}/controls/merchants/block`, {
      method: 'POST',
      body: JSON.stringify({ category }),
    });
  }

  /**
   * Unblock a merchant category
   */
  async unblockMerchantCategory(cardId: string, category: string): Promise<void> {
    await this.request(`/cards/${cardId}/controls/merchants/unblock`, {
      method: 'POST',
      body: JSON.stringify({ category }),
    });
  }

  /**
   * Block a country
   */
  async blockCountry(cardId: string, countryCode: string): Promise<void> {
    await this.request(`/cards/${cardId}/controls/geo/block`, {
      method: 'POST',
      body: JSON.stringify({ country_code: countryCode }),
    });
  }

  /**
   * Unblock a country
   */
  async unblockCountry(cardId: string, countryCode: string): Promise<void> {
    await this.request(`/cards/${cardId}/controls/geo/unblock`, {
      method: 'POST',
      body: JSON.stringify({ country_code: countryCode }),
    });
  }

  // ===========================================================================
  // PIN MANAGEMENT
  // ===========================================================================

  /**
   * Set initial PIN
   */
  async setPin(cardId: string, pin: string): Promise<void> {
    await this.request(`/cards/${cardId}/pin/set`, {
      method: 'POST',
      body: JSON.stringify({ pin }),
    });
  }

  /**
   * Change existing PIN
   */
  async changePin(cardId: string, currentPin: string, newPin: string): Promise<void> {
    await this.request(`/cards/${cardId}/pin/change`, {
      method: 'PUT',
      body: JSON.stringify({
        current_pin: currentPin,
        new_pin: newPin,
      }),
    });
  }

  /**
   * Get PIN status
   */
  async getPinStatus(cardId: string): Promise<{
    pinSet: boolean;
    pinLocked: boolean;
    failedAttempts: number;
  }> {
    return this.request(`/cards/${cardId}/pin/status`);
  }

  /**
   * Request OTP for PIN reset
   */
  async requestPinResetOtp(cardId: string): Promise<void> {
    await this.request(`/cards/${cardId}/pin/reset/otp`, {
      method: 'POST',
    });
  }

  /**
   * Reset PIN with OTP verification
   */
  async resetPin(cardId: string, otp: string, newPin: string): Promise<void> {
    await this.request(`/cards/${cardId}/pin/reset`, {
      method: 'POST',
      body: JSON.stringify({
        verification_method: 'otp',
        verification_data: { otp_code: otp },
        new_pin: newPin,
      }),
    });
  }

  /**
   * Unlock PIN after failed attempts
   */
  async unlockPin(cardId: string): Promise<void> {
    await this.request(`/cards/${cardId}/pin/unlock`, {
      method: 'POST',
    });
  }

  // ===========================================================================
  // TRANSACTIONS
  // ===========================================================================

  /**
   * Get transactions for a card
   */
  async getTransactions(
    cardId: string,
    options?: {
      startDate?: string;
      endDate?: string;
      minAmount?: number;
      maxAmount?: number;
      transactionType?: string;
      status?: string;
      merchantName?: string;
      category?: string;
      page?: number;
      pageSize?: number;
    }
  ): Promise<{ transactions: FISTransaction[]; total: number; page: number }> {
    const params = new URLSearchParams();
    if (options?.startDate) params.append('start_date', options.startDate);
    if (options?.endDate) params.append('end_date', options.endDate);
    if (options?.minAmount) params.append('min_amount', options.minAmount.toString());
    if (options?.maxAmount) params.append('max_amount', options.maxAmount.toString());
    if (options?.transactionType) params.append('transaction_type', options.transactionType);
    if (options?.status) params.append('status', options.status);
    if (options?.merchantName) params.append('merchant_name', options.merchantName);
    if (options?.category) params.append('category', options.category);
    if (options?.page) params.append('page', options.page.toString());
    if (options?.pageSize) params.append('page_size', options.pageSize.toString());

    return this.request(`/cards/${cardId}/transactions?${params.toString()}`);
  }

  /**
   * Get recent transactions
   */
  async getRecentTransactions(cardId: string, limit: number = 10): Promise<FISTransaction[]> {
    return this.request(`/cards/${cardId}/transactions/recent?limit=${limit}`);
  }

  /**
   * Get pending transactions
   */
  async getPendingTransactions(cardId: string): Promise<FISTransaction[]> {
    return this.request(`/cards/${cardId}/transactions/pending`);
  }

  /**
   * Get transaction details
   */
  async getTransaction(cardId: string, transactionId: string): Promise<FISTransaction> {
    return this.request(`/cards/${cardId}/transactions/${transactionId}`);
  }

  /**
   * Get spending summary
   */
  async getSpendingSummary(
    cardId: string,
    startDate: string,
    endDate: string
  ): Promise<{
    totalSpent: number;
    totalRefunds: number;
    transactionCount: number;
    averageTransaction: number;
  }> {
    return this.request(
      `/cards/${cardId}/transactions/summary?start_date=${startDate}&end_date=${endDate}`
    );
  }

  /**
   * Get spending by category
   */
  async getSpendingByCategory(
    cardId: string,
    startDate: string,
    endDate: string
  ): Promise<Array<{ category: string; amount: number; count: number }>> {
    return this.request(
      `/cards/${cardId}/transactions/categories?start_date=${startDate}&end_date=${endDate}`
    );
  }

  /**
   * Initiate a dispute
   */
  async initiateDispute(
    cardId: string,
    transactionId: string,
    reason: string,
    description: string
  ): Promise<{ disputeId: string; status: string }> {
    return this.request(`/cards/${cardId}/transactions/${transactionId}/dispute`, {
      method: 'POST',
      body: JSON.stringify({ reason, description }),
    });
  }

  // ===========================================================================
  // FRAUD & SECURITY
  // ===========================================================================

  /**
   * Get security alerts
   */
  async getAlerts(options?: {
    cardId?: string;
    status?: string;
    priority?: string;
  }): Promise<FISAlert[]> {
    const params = new URLSearchParams();
    if (options?.cardId) params.append('card_id', options.cardId);
    if (options?.status) params.append('status', options.status);
    if (options?.priority) params.append('priority', options.priority);

    return this.request(`/alerts?${params.toString()}`);
  }

  /**
   * Get unread alert count
   */
  async getUnreadAlertCount(cardId?: string): Promise<number> {
    const params = cardId ? `?card_id=${cardId}` : '';
    const data = await this.request<{ count: number }>(`/alerts/unread/count${params}`);
    return data.count;
  }

  /**
   * Acknowledge an alert
   */
  async acknowledgeAlert(alertId: string): Promise<void> {
    await this.request(`/alerts/${alertId}/acknowledge`, {
      method: 'PUT',
    });
  }

  /**
   * Resolve an alert
   */
  async resolveAlert(alertId: string, resolution: string): Promise<void> {
    await this.request(`/alerts/${alertId}/resolve`, {
      method: 'PUT',
      body: JSON.stringify({ resolution, is_false_positive: false }),
    });
  }

  /**
   * Report fraud
   */
  async reportFraud(
    cardId: string,
    fraudType: string,
    description: string,
    transactionId?: string
  ): Promise<{ reportId: string }> {
    return this.request('/fraud/reports', {
      method: 'POST',
      body: JSON.stringify({
        card_id: cardId,
        fraud_type: fraudType,
        description,
        transaction_id: transactionId,
      }),
    });
  }

  /**
   * Set travel notice
   */
  async setTravelNotice(
    cardId: string,
    startDate: string,
    endDate: string,
    destinations: string[],
    notes?: string
  ): Promise<TravelNotice> {
    return this.request(`/cards/${cardId}/travel-notices`, {
      method: 'POST',
      body: JSON.stringify({
        start_date: startDate,
        end_date: endDate,
        destinations,
        notes,
      }),
    });
  }

  /**
   * Get travel notices
   */
  async getTravelNotices(cardId: string): Promise<TravelNotice[]> {
    return this.request(`/cards/${cardId}/travel-notices`);
  }

  /**
   * Cancel travel notice
   */
  async cancelTravelNotice(cardId: string, noticeId: string): Promise<void> {
    await this.request(`/cards/${cardId}/travel-notices/${noticeId}`, {
      method: 'DELETE',
    });
  }

  // ===========================================================================
  // DIGITAL WALLETS
  // ===========================================================================

  /**
   * Check Apple Pay eligibility
   */
  async checkApplePayEligibility(cardId: string): Promise<{ eligible: boolean; reason?: string }> {
    return this.request(`/cards/${cardId}/wallet/apple-pay/eligibility`);
  }

  /**
   * Check Google Pay eligibility
   */
  async checkGooglePayEligibility(cardId: string): Promise<{ eligible: boolean; reason?: string }> {
    return this.request(`/cards/${cardId}/wallet/google-pay/eligibility`);
  }

  /**
   * Get wallet tokens
   */
  async getWalletTokens(cardId: string): Promise<FISWalletToken[]> {
    return this.request(`/cards/${cardId}/wallet/tokens`);
  }

  /**
   * Suspend a wallet token
   */
  async suspendWalletToken(cardId: string, tokenId: string): Promise<void> {
    await this.request(`/cards/${cardId}/wallet/tokens/${tokenId}/suspend`, {
      method: 'POST',
    });
  }

  /**
   * Resume a wallet token
   */
  async resumeWalletToken(cardId: string, tokenId: string): Promise<void> {
    await this.request(`/cards/${cardId}/wallet/tokens/${tokenId}/resume`, {
      method: 'POST',
    });
  }

  /**
   * Delete a wallet token
   */
  async deleteWalletToken(cardId: string, tokenId: string): Promise<void> {
    await this.request(`/cards/${cardId}/wallet/tokens/${tokenId}`, {
      method: 'DELETE',
    });
  }

  /**
   * Suspend all wallet tokens (for lost/stolen card)
   */
  async suspendAllWalletTokens(cardId: string): Promise<void> {
    await this.request(`/cards/${cardId}/wallet/tokens/suspend-all`, {
      method: 'POST',
    });
  }
}

// Export singleton instance
export const fisCardService = new FISCardService();
export default fisCardService;

File: swipesavvy-ai-agents/app/services/fis_card_service.py
"""
FIS Global Payment One - Card Issuance Service

Handles card lifecycle operations:
- Virtual card issuance
- Physical card ordering
- Card activation
- Card replacement
- Card cancellation
"""

import logging
from typing import Optional, Dict, Any, List
from datetime import datetime
from uuid import UUID
from pydantic import BaseModel, Field

from app.services.fis_global_service import (
    FISGlobalService,
    FISAPIResponse,
    FISCardStatus,
    FISCardType,
    get_fis_service
)

logger = logging.getLogger(__name__)


# =============================================================================
# REQUEST/RESPONSE MODELS
# =============================================================================

class IssueVirtualCardRequest(BaseModel):
    """Request to issue a virtual card"""
    user_id: UUID
    cardholder_name: str
    nickname: Optional[str] = None
    set_as_primary: bool = False


class IssuePhysicalCardRequest(BaseModel):
    """Request to order a physical card"""
    user_id: UUID
    cardholder_name: str
    shipping_address: Dict[str, str]  # street, city, state, zip, country
    expedited: bool = False
    nickname: Optional[str] = None
    set_as_primary: bool = False


class ActivateCardRequest(BaseModel):
    """Request to activate a card"""
    card_id: str
    last_four: str  # For verification
    activation_code: Optional[str] = None  # For physical cards


class ReplaceCardRequest(BaseModel):
    """Request to replace a card"""
    card_id: str
    reason: str  # lost, stolen, damaged, expired
    shipping_address: Optional[Dict[str, str]] = None  # For physical cards
    expedited: bool = False


class CardResponse(BaseModel):
    """Card response data"""
    card_id: str
    fis_card_id: str
    card_type: str
    status: str
    last_four: str
    expiry_month: int
    expiry_year: int
    cardholder_name: str
    card_network: str = "visa"
    is_primary: bool = False
    nickname: Optional[str] = None
    created_at: datetime
    activated_at: Optional[datetime] = None


# =============================================================================
# FIS CARD SERVICE
# =============================================================================

class FISCardService:
    """
    Service for FIS card issuance and lifecycle management.
    """

    def __init__(self, fis_service: Optional[FISGlobalService] = None):
        """
        Initialize card service.

        Args:
            fis_service: FIS Global service instance (uses singleton if not provided)
        """
        self.fis = fis_service or get_fis_service()

    async def issue_virtual_card(
        self,
        user_id: UUID,
        cardholder_name: str,
        nickname: Optional[str] = None,
        set_as_primary: bool = False
    ) -> FISAPIResponse:
        """
        Issue a new virtual card for instant use.

        Args:
            user_id: User's UUID
            cardholder_name: Name to emboss on card
            nickname: Optional friendly name for the card
            set_as_primary: Whether to set as primary card

        Returns:
            FISAPIResponse with card details
        """
        logger.info(f"Issuing virtual card for user {user_id}")

        payload = {
            "card_type": "virtual",
            "cardholder_name": cardholder_name.upper(),
            "external_user_id": str(user_id),
            "program_id": "swipesavvy_virtual",
            "metadata": {
                "nickname": nickname,
                "is_primary": set_as_primary
            }
        }

        response = await self.fis._make_request(
            method="POST",
            endpoint="/cards/issue",
            data=payload
        )

        if response.success:
            logger.info(f"Virtual card issued successfully: {response.data.get('card_id')}")
        else:
            logger.error(f"Failed to issue virtual card: {response.error_message}")

        return response

    async def order_physical_card(
        self,
        user_id: UUID,
        cardholder_name: str,
        shipping_address: Dict[str, str],
        expedited: bool = False,
        nickname: Optional[str] = None,
        set_as_primary: bool = False
    ) -> FISAPIResponse:
        """
        Order a physical card for mailing.

        Args:
            user_id: User's UUID
            cardholder_name: Name to emboss on card
            shipping_address: Shipping address dict with street, city, state, zip, country
            expedited: Whether to use expedited shipping
            nickname: Optional friendly name for the card
            set_as_primary: Whether to set as primary card

        Returns:
            FISAPIResponse with card and shipping details
        """
        logger.info(f"Ordering physical card for user {user_id}")

        payload = {
            "card_type": "physical",
            "cardholder_name": cardholder_name.upper(),
            "external_user_id": str(user_id),
            "program_id": "swipesavvy_physical",
            "shipping": {
                "address": {
                    "line1": shipping_address.get("street"),
                    "city": shipping_address.get("city"),
                    "state": shipping_address.get("state"),
                    "postal_code": shipping_address.get("zip"),
                    "country": shipping_address.get("country", "US")
                },
                "expedited": expedited,
                "method": "expedited" if expedited else "standard"
            },
            "metadata": {
                "nickname": nickname,
                "is_primary": set_as_primary
            }
        }

        response = await self.fis._make_request(
            method="POST",
            endpoint="/cards/issue",
            data=payload
        )

        if response.success:
            logger.info(f"Physical card ordered successfully: {response.data.get('card_id')}")
        else:
            logger.error(f"Failed to order physical card: {response.error_message}")

        return response

    async def activate_card(
        self,
        card_id: str,
        last_four: str,
        activation_code: Optional[str] = None
    ) -> FISAPIResponse:
        """
        Activate a card.

        Args:
            card_id: FIS card ID
            last_four: Last 4 digits for verification
            activation_code: Activation code (for physical cards)

        Returns:
            FISAPIResponse with activation status
        """
        logger.info(f"Activating card {card_id}")

        payload = {
            "last_four_verification": last_four
        }

        if activation_code:
            payload["activation_code"] = activation_code

        response = await self.fis._make_request(
            method="POST",
            endpoint=f"/cards/{card_id}/activate",
            data=payload
        )

        if response.success:
            logger.info(f"Card {card_id} activated successfully")
        else:
            logger.error(f"Failed to activate card {card_id}: {response.error_message}")

        return response

    async def get_card(self, card_id: str) -> FISAPIResponse:
        """
        Get card details.

        Args:
            card_id: FIS card ID

        Returns:
            FISAPIResponse with card details
        """
        return await self.fis._make_request(
            method="GET",
            endpoint=f"/cards/{card_id}"
        )

    async def get_user_cards(self, user_id: UUID) -> FISAPIResponse:
        """
        Get all cards for a user.

        Args:
            user_id: User's UUID

        Returns:
            FISAPIResponse with list of cards
        """
        return await self.fis._make_request(
            method="GET",
            endpoint="/cards",
            params={"external_user_id": str(user_id)}
        )

    async def get_card_sensitive_data(
        self,
        card_id: str,
        include_pan: bool = False,
        include_cvv: bool = False
    ) -> FISAPIResponse:
        """
        Get sensitive card data (PAN, CVV) - requires additional authorization.

        Args:
            card_id: FIS card ID
            include_pan: Whether to include full PAN
            include_cvv: Whether to include CVV

        Returns:
            FISAPIResponse with sensitive card data
        """
        logger.info(f"Retrieving sensitive data for card {card_id}")

        params = {}
        if include_pan:
            params["include_pan"] = "true"
        if include_cvv:
            params["include_cvv"] = "true"

        return await self.fis._make_request(
            method="GET",
            endpoint=f"/cards/{card_id}/sensitive",
            params=params
        )

    async def replace_card(
        self,
        card_id: str,
        reason: str,
        shipping_address: Optional[Dict[str, str]] = None,
        expedited: bool = False
    ) -> FISAPIResponse:
        """
        Replace a card (lost, stolen, damaged, expired).

        Args:
            card_id: FIS card ID of card to replace
            reason: Reason for replacement (lost, stolen, damaged, expired)
            shipping_address: New shipping address (uses existing if not provided)
            expedited: Whether to use expedited shipping

        Returns:
            FISAPIResponse with new card details
        """
        logger.info(f"Replacing card {card_id} - reason: {reason}")

        payload = {
            "reason": reason,
            "expedited": expedited
        }

        if shipping_address:
            payload["shipping"] = {
                "address": {
                    "line1": shipping_address.get("street"),
                    "city": shipping_address.get("city"),
                    "state": shipping_address.get("state"),
                    "postal_code": shipping_address.get("zip"),
                    "country": shipping_address.get("country", "US")
                }
            }

        response = await self.fis._make_request(
            method="POST",
            endpoint=f"/cards/{card_id}/replace",
            data=payload
        )

        if response.success:
            logger.info(f"Card {card_id} replaced - new card: {response.data.get('new_card_id')}")
        else:
            logger.error(f"Failed to replace card {card_id}: {response.error_message}")

        return response

    async def cancel_card(
        self,
        card_id: str,
        reason: str
    ) -> FISAPIResponse:
        """
        Cancel/close a card permanently.

        Args:
            card_id: FIS card ID
            reason: Reason for cancellation

        Returns:
            FISAPIResponse with cancellation status
        """
        logger.info(f"Cancelling card {card_id} - reason: {reason}")

        response = await self.fis._make_request(
            method="DELETE",
            endpoint=f"/cards/{card_id}",
            data={"reason": reason}
        )

        if response.success:
            logger.info(f"Card {card_id} cancelled successfully")
        else:
            logger.error(f"Failed to cancel card {card_id}: {response.error_message}")

        return response

    async def update_card_status(
        self,
        card_id: str,
        status: FISCardStatus
    ) -> FISAPIResponse:
        """
        Update card status.

        Args:
            card_id: FIS card ID
            status: New card status

        Returns:
            FISAPIResponse with updated status
        """
        logger.info(f"Updating card {card_id} status to {status.value}")

        return await self.fis._make_request(
            method="PUT",
            endpoint=f"/cards/{card_id}/status",
            data={"status": status.value}
        )

    async def get_shipping_status(self, card_id: str) -> FISAPIResponse:
        """
        Get shipping status for a physical card.

        Args:
            card_id: FIS card ID

        Returns:
            FISAPIResponse with shipping details
        """
        return await self.fis._make_request(
            method="GET",
            endpoint=f"/cards/{card_id}/shipping"
        )


# =============================================================================
# SINGLETON INSTANCE
# =============================================================================

_card_service: Optional[FISCardService] = None


def get_fis_card_service() -> FISCardService:
    """Get or create singleton FIS Card service instance"""
    global _card_service
    if _card_service is None:
        _card_service = FISCardService()
    return _card_service


# Export singleton
fis_card_service = get_fis_card_service()
